In [6]:
import tensorflow as tf
tf.__version__

'2.19.0'

In [14]:
print("GPU's Available: ", len(tf.config.experimental.list_physical_devices('GPU')))

GPU's Available:  1


# <font color=red>Mr Fugu Data Science</font>

# `Installing TensorFlow on Apple Silicon 3 Different Options:`


# *Download/Install Xcode:* `xcode-select --install`

# `Conda forge vs Anaconda for ML models Mac:`

https://conda-forge.org/docs/user/introduction/

Install Mini-conda (change default channels) for your installed packages (to avoid commercial license problems)

Read the reasons you may need this from the citation at bottom of document.

or

Mini-Forge Installer

https://www.anaconda.com/blog/is-conda-free`


# **Here is something interesting:**

**`Depending on how you download and install TensorFlow on device gives vastly different versions.`**

`I have 2 ways I installed this on Apple Silicon M4 today in 2 different Environments:`

`--------------------------------------------------`

# 1.) (Older Versions `2.12<`)

+ `conda create --name TF_Env --no-default-packages`

+ `conda activate TF_Env`

+ `conda install -c apple tensorflow-deps`

+ `conda install tensorflow` (This will work but downgrades packages because of last step dependencies)

+ `pip install tensorflow-metal`

+ `pip install ipykernel`

+ switch to location where Jupyter Notebook is located, usually (base): `conda activate base` then run `jupyter notebook`

+ in the GUI go to `New` and you will see the updated kernel for `TF_Env`

+ `import tensorflow as tf`, `tf.__version__`

(gives me 2.12)

+ `print("GPU's Available: ", len(tf.config.experimental.list_physical_devices('GPU')))`


+ EX.)

`cifar = tf.keras.datasets.cifar100
(x_train, y_train), (x_test, y_test) = cifar.load_data()
model = tf.keras.applications.ResNet50(
    include_top=True,
    weights=None,
    input_shape=(32, 32, 3),
    classes=100,)
loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=False)
model.compile(optimizer="adam", loss=loss_fn, metrics=["accuracy"])
model.fit(x_train, y_train, epochs=5, batch_size=64)`

`--------------------------------------------------`

# 2.) (Never Version `2.13 >=`)

+ `conda create --name TF_Env_V2 --no-default-packages`

+ `conda activate TF_Env_V2 `

+ `conda install pip`

+ `pip install tensorflow` (you will get an error! this is due to Python Versions not matching with TF)

+ Then you need to downgrade Python because I had 13 install :(

`conda install python==3.11`

+ `pip install tensorflow` now it will work!

+ `pip install tensorflow-metal`

+ `pip install ipykernel`

+ Now `conda activate base` and then when it opens `jupyter notebook`
    + Check your kernels in the browser and it will be updated
    + Next `import tensorflow as tf`, `tf.__version__` and I have 2.19
      
+ `print("GPU's Available: ", len(tf.config.experimental.list_physical_devices('GPU')))`


+ EX.)

`cifar = tf.keras.datasets.cifar100
(x_train, y_train), (x_test, y_test) = cifar.load_data()
model = tf.keras.applications.ResNet50(
    include_top=True,
    weights=None,
    input_shape=(32, 32, 3),
    classes=100,)
loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=False)
model.compile(optimizer="adam", loss=loss_fn, metrics=["accuracy"])
model.fit(x_train, y_train, epochs=5, batch_size=64)`

`--------------------------------------------------`

**This Section for those using TensorFlow outside of a Conda Env**


# 3.) ` This is for your vm/editor` 

+ **Create an environment:**


  `python3 -m venv .venv`
  
+ **Activate the .venv for use:**

 `source .venv/bin/activate`
 
+ **Need to Install Pip/upgrade version:**

`pip install -U pip`

+ `Install TensorFlow and GPU usuage (metal)`
  
 `pip install tensorflow tensorflow-metal`

+ **Open Python or your Editor and check that it worked:**

`import tensorflow as tf`, and this afterward `tf.__version__`

Ex.) `cifar = tf.keras.datasets.cifar100
(x_train, y_train), (x_test, y_test) = cifar.load_data()
model = tf.keras.applications.ResNet50(
    include_top=True,
    weights=None,
    input_shape=(32, 32, 3),
    classes=100,)
loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=False)
model.compile(optimizer="adam", loss=loss_fn, metrics=["accuracy"])
model.fit(x_train, y_train, epochs=5, batch_size=64)`

# **`Install JAX:`**

In [32]:
!pip install jax-metal

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 41.2/41.2 MB 26.7 MB/s eta 0:00:00a 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.3/2.3 MB 45.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 53.5/53.5 MB 55.8 MB/s eta 0:00:00a 0:00:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 22.4/22.4 MB 54.4 MB/s eta 0:00:0000:0100:01


In [ ]:
import jax

In [36]:
jax.__version__

'0.6.0'

In [40]:
print(jax.numpy.arange(10))

[0 1 2 3 4 5 6 7 8 9]


In [46]:
!pip install -U jaxlib jax

In [ ]:
# `ENABLE_PJRT_COMPATIBILITY=1` this may be need in some case... if you get errors.